airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultDiscoverCatalogWorker.java

```diff
@@ -73,43 +70,40 @@ public ConnectorJobOutput run(final StandardDiscoverCatalogInput discoverSchemaI
           jobRoot,
           WorkerConstants.SOURCE_CONFIG_JSON_FILENAME,
           Jsons.serialize(discoverSchemaInput.getConnectionConfiguration()));
+      final int exitCode = process.exitValue();
+      final String exitCodeMessage = String.format("Discover job subprocess finished with exit code %s", exitCode);
+      LOGGER.debug(exitCodeMessage);
+      ConnectorJobOutput jobOutput = new ConnectorJobOutput().withOutputType(OutputType.DISCOVER_CATALOG_ID);
 
       LineGobbler.gobble(process.getErrorStream(), LOGGER::error);
 
-      final Map<Type, List<AirbyteMessage>> messagesByType;
-
-      try (final InputStream stdout = process.getInputStream()) {
-        messagesByType = streamFactory.create(IOs.newBufferedReader(stdout))
-            .collect(Collectors.groupingBy(AirbyteMessage::getType));
-
-        WorkerUtils.gentleClose(process, 30, TimeUnit.MINUTES);
-      }
+      final Map<Type, List<AirbyteMessage>> messagesByType = WorkerUtils.getMessagesByType(process, streamFactory, 30);
 
       final Optional<AirbyteCatalog> catalog = messagesByType
           .getOrDefault(Type.CATALOG, new ArrayList<>()).stream()
           .map(AirbyteMessage::getCatalog)
           .findFirst();
 
-      final int exitCode = process.exitValue();
-      if (exitCode == 0) {
-        if (catalog.isEmpty()) {
-          throw new WorkerException("Integration failed to output a catalog struct.");
-        }
+      final Optional<FailureReason> failureReason = WorkerUtils.getJobFailureReasonFromMessages(OutputType.DISCOVER_CATALOG_ID, messagesByType);
+      if (failureReason.isPresent()) {
+        jobOutput = jobOutput.withFailureReason(failureReason.get());
+      }
 
+      if (catalog.isPresent()) {
         final UUID catalogId =
             configRepository.writeActorCatalogFetchEvent(catalog.get(),
                 // NOTE: sourceId is marked required in the OpenAPI config but the code generator doesn't enforce
                 // it, so we check again here.
                 discoverSchemaInput.getSourceId() == null ? null : UUID.fromString(discoverSchemaInput.getSourceId()),
                 discoverSchemaInput.getConnectorVersion(),
                 discoverSchemaInput.getConfigHash());
-        return new ConnectorJobOutput().withOutputType(OutputType.DISCOVER_CATALOG_ID).withDiscoverCatalogId(catalogId);
+        jobOutput = jobOutput.withDiscoverCatalogId(catalogId);
       } else {
-        return WorkerUtils.getJobFailureOutputOrThrow(
-            OutputType.DISCOVER_CATALOG_ID,
-            messagesByType,
-            String.format("Discover job subprocess finished with exit code %s", exitCode));
+        if (failureReason.isEmpty()) {
+          throw new WorkerException("Integration failed to output a catalog struct.");
+        }
```

Comment:
This means that if a connector:
- exited with a non-zero exit code
-  _did_ output a catalog
- _did not_ output a failure reason

We would consider this a success. Should these conditions throw an exception as well? A non-zero exit code probably means _something_ went wrong but we just don't know what (since there isn't a failure reason)

Reply:
Yes, you are right. 
But should we throw an exception if we can parse a catalog?  
I'd like to characterize an error only when we can't retrieve what we are looking for (spec, connection status, catalog).

Reply:
I don't feel super strongly about this, but I think having a non-zero exit code returned by the source means something went wrong so the job should fail. My worry would be that even though we have a catalog, a non-zero exit code may mean we shouldn't trust that catalog to be correct. Ideally we get a failure reason that tells us more, but it's kind of ambiguous if there isn't one present.

I can see the other side of the argument though. The protocol doesn't really say anything about the exit code, so just the presence of the right type of message might be enough on that side. I can also see it being ok to process the output if the exit code indicates something like the container was killed because it ran out of memory.

Reply:
FWIW, I'd definitely be confused if I saw a log that said a job exited with a non-zero status, and everything continued as planned. If we think we can generally trust the catalog outputted in this case, at least indicating a non-zero exit code as a warning might be helpful for hunting it down if things get messed up due to a strange catalog?

Reply:
@erohmensing I've changed the logic to log a warning if status code != 0.